### PR TITLE
Modifying some app functionality

### PR DESCRIPTION
**Pull request by maraujop**
I have modified the storage and model so that read and unread messages are both pulled from the DB, and using CSS properties the user can set up how to display the messages. The user could make unread messages not visible for example. 

If you don't consider this necessary, I could put everything apart in an all_messages method that returns all messages like this. {{ messages.all_messages }} Or maybe add a context_processor and a variable like {{ all_messages }}

I have also added a number_unread method that let's the programmer show the user how many messages are left to read. I find this a must feature in any app I can think of. {{ messages.number_unread }}

Maybe README should be adapted if you pull this changeset. Also moving "read" and "unread" tags to I18N might be interesting or moved to constants. 

Thanks :)

### CODE REVIEW

**Comment by samluescher** [comment]
Hi! Thanks for the contribution. While I like the idea of the read/unread tag, I think your decision to fetch all messages (read and unread) in the _get() method changes the nature/purpose of the app too much. While it could be used for a fully featured messaging system (also between users), it really is just an extension of Django's original messages app, the purpose of which is to create system messages and make sure they are displayed. Additionally, this app just makes sure that important messages actually are noticed.

What you are building, on the other hand, is more of a message inbox. I think it's a good idea to add an extra context processor called `all_messages` that looks just like django.contrib.messages.context_processors.messages and that does what you propose. For a clean design, I'd add an `exclude_read=False` (default is False!) parameter to the _get() method of the storage, and add a method get_all_messages() to the api, and the context processor would look like this:

```
from persistent_messages.api import get_all_messages


def messages(request):
    """
    Returns a lazy 'all_messages' context variable, containing all (read and unread) messages, 
    as opposed to django.contrib.messages.context_processors.messages that just returns unread messages.
    """
    return {'all_messages': get_all_messages(request)}
```

Note that this should be LAZY.

Finally, what would be really useful in my opinion is adding a `message_inbox` view to views.py. That would serve the purpose much better, I think. After all, there already is a `message_detail` view. This view should render a queryset containing all read+unread messages (maybe with optional filter params?) thorugh a table template, linking them to the detail views.

**Comment by maraujop** [comment]
Hi philomat! 

I've recently started using your Django app and I find your approach to persistent messages very smart and neat. So thank you for releasing the code to the public :)

Now let's get down to business.

> While I like the idea of the read/unread tag, I think your decision to fetch all messages (read and unread) in the _get() method changes the nature/purpose of the app too much. 

I have to say, I really thought about it and as you could notice, I wasn't sure my approach was the best. That's why I suggested moving the code elsewhere. 

> While it could be used for a fully featured messaging system (also between users), it really is just an extension of Django's original messages app, the purpose of which is to create system messages and make sure they are displayed. Additionally, this app just makes sure that important messages actually are noticed.

I totally understand. I have to say I don't really agree with Django's messages approach, I think it should be more extensible. When I saw that in your model you handle a `from_user` and `user` I thought you were interested in extending a little bit messages. 

> What you are building, on the other hand, is more of a message inbox. I think it's a good idea to add an extra context processor called all_messages that looks just like django.contrib.messages.context_processors.messages and that does what you propose. For a clean design, I'd add an exclude_read=False (default is False!) parameter to the _get() method of the storage, and add a method get_all_messages() to the api, and the context processor would look like this.

As the `_get` method is called by Django messages, I guess the right way to go would be to have a settings variable that allow the user decide how to filter. Maybe something like `MESSAGE_EXCLUDE_READ`

This way we could use messages built in context processor, without defining a new context processor.

> Note that this should be LAZY.

<b>Very important!</b>

> Finally, what would be really useful in my opinion is adding a message_inbox view to views.py. That would serve the purpose much better, I think. After all, there already is a message_detail view. This view should render a queryset containing all read+unread messages (maybe with optional filter params?) thorugh a table template, linking them to the detail views.

I thought about this option too. Damn! there are many alternatives: context processor, settings variable, extra method to messages and a view. 

I think having a date range might be interesting in these optional filter params, what do you think?

So to sum up a little bit, I think that adding your Django app the possibility of working as an inter user messaging system (I have to clarify I'm not using the system like that) is a very interesting feature. That way you let the user choose how to use it.

Adding a `delete_message` view is also something I am considering. Let me know if you want me to code it and add an AJAX connection. You can count on me for coding :D

Cheers

**Comment by maraujop** [comment]
I'm coming up with more changes we could make to the application. Let me know, how you want to discuss them :)

**Comment by maraujop** [comment]
Hi philomat,

I've seen you have used one of my commits, but I'm still waiting for an answer on this issue or pull request. 

Let me know if you want to amplify your app functionality to be an inter-user message system or you would like to keep it as it is. Because otherwise, I will keep working on my fork, adding what I think it's missing. I need to move this a little bit, time is running out for me :)

Thanks, cheers
